// @gala-chain/client/lib/src/generic/index.js
"use strict";
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadJson = exports.createRandomKeys = exports.ChainUser = exports.ChainClientBuilder = exports.isClassType = exports.ChainClient = void 0;
const tslib_1 = require("tslib");
var ChainClient_1 = require("./ChainClient");
Object.defineProperty(exports, "ChainClient", { enumerable: true, get: function () { return ChainClient_1.ChainClient; } });
Object.defineProperty(exports, "isClassType", { enumerable: true, get: function () { return ChainClient_1.isClassType; } });
var ChainClientBuilder_1 = require("./ChainClientBuilder");
Object.defineProperty(exports, "ChainClientBuilder", { enumerable: true, get: function () { return ChainClientBuilder_1.ChainClientBuilder; } });
var ChainUser_1 = require("./ChainUser");
Object.defineProperty(exports, "ChainUser", { enumerable: true, get: function () { return ChainUser_1.ChainUser; } });
exports.createRandomKeys = tslib_1.__importStar(require("./createRandomKeys"));
var loadJson_1 = require("./loadJson");
Object.defineProperty(exports, "loadJson", { enumerable: true, get: function () { return loadJson_1.loadJson; } });
