// @gala-chain/api/lib/src/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signatures = exports.getValidationErrorInfo = exports.generateResponseSchema = exports.generateSchema = exports.serialize = exports.deserialize = void 0;
const tslib_1 = require("tslib");
const deserialize_1 = tslib_1.__importDefault(require("./deserialize"));
exports.deserialize = deserialize_1.default;
const generate_schema_1 = require("./generate-schema");
Object.defineProperty(exports, "generateResponseSchema", { enumerable: true, get: function () { return generate_schema_1.generateResponseSchema; } });
Object.defineProperty(exports, "generateSchema", { enumerable: true, get: function () { return generate_schema_1.generateSchema; } });
const getValidationErrorMessage_1 = require("./getValidationErrorMessage");
Object.defineProperty(exports, "getValidationErrorInfo", { enumerable: true, get: function () { return getValidationErrorMessage_1.getValidationErrorInfo; } });
const serialize_1 = tslib_1.__importDefault(require("./serialize"));
exports.serialize = serialize_1.default;
const signatures_1 = tslib_1.__importDefault(require("./signatures"));
exports.signatures = signatures_1.default;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
tslib_1.__exportStar(require("./chain-decorators"), exports);
tslib_1.__exportStar(require("./error"), exports);
tslib_1.__exportStar(require("./transform-decorators"), exports);
