// @gala-chain/chaincode/lib/src/locks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidExpirationError = exports.UnlockForbiddenUserError = exports.NftInvalidQuantityLockError = exports.MissingLockError = exports.LockTokenFailedError = exports.unlockTokens = exports.lockTokens = exports.lockToken = exports.unlockToken = void 0;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const LockError_1 = require("./LockError");
Object.defineProperty(exports, "InvalidExpirationError", { enumerable: true, get: function () { return LockError_1.InvalidExpirationError; } });
Object.defineProperty(exports, "LockTokenFailedError", { enumerable: true, get: function () { return LockError_1.LockTokenFailedError; } });
Object.defineProperty(exports, "MissingLockError", { enumerable: true, get: function () { return LockError_1.MissingLockError; } });
Object.defineProperty(exports, "NftInvalidQuantityLockError", { enumerable: true, get: function () { return LockError_1.NftInvalidQuantityLockError; } });
Object.defineProperty(exports, "UnlockForbiddenUserError", { enumerable: true, get: function () { return LockError_1.UnlockForbiddenUserError; } });
const lockTokens_1 = require("./lockTokens");
Object.defineProperty(exports, "lockToken", { enumerable: true, get: function () { return lockTokens_1.lockToken; } });
Object.defineProperty(exports, "lockTokens", { enumerable: true, get: function () { return lockTokens_1.lockTokens; } });
const unlockToken_1 = require("./unlockToken");
Object.defineProperty(exports, "unlockToken", { enumerable: true, get: function () { return unlockToken_1.unlockToken; } });
Object.defineProperty(exports, "unlockTokens", { enumerable: true, get: function () { return unlockToken_1.unlockTokens; } });
