// @gala-chain/chaincode/lib/src/contracts/GalaContract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GalaContract = void 0;
const tslib_1 = require("tslib");
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const api_1 = require("@gala-chain/api");
const fabric_contract_api_1 = require("fabric-contract-api");
const types_1 = require("../types");
const utils_1 = require("../utils");
const GalaContractApi_1 = require("./GalaContractApi");
const GalaTransaction_1 = require("./GalaTransaction");
const tracing_1 = require("./tracing");
class GalaContract extends fabric_contract_api_1.Contract {
    /**
     * @param name Contract name
     * @param version Contract version. The actual value should be defined in the child
     *    * class, and should be taken from package.json. If you extend contract class
     *    * that extends GalaContract, you should also override version.
     */
    constructor(name, version) {
        super(name);
        this.version = version;
    }
    getVersion() {
        return this.version;
    }
    createContext() {
        return new types_1.GalaChainContext();
    }
    async beforeTransaction(ctx) {
        await (0, tracing_1.trace)("before", ctx, () => super.beforeTransaction(ctx));
    }
    // eslint-disable-next-line @typescript-eslint/ban-types
    async aroundTransaction(ctx, fn, parameters) {
        // note: Fabric uses Promise<void> type, but actually it returns transaction result
        return await (0, tracing_1.trace)("around", ctx, () => super.aroundTransaction(ctx, fn, parameters));
    }
    async afterTransaction(ctx, result) {
        await (0, tracing_1.trace)("after", ctx, async () => {
            var _a, _b, _c;
            await super.afterTransaction(ctx, result);
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            if (typeof result === "object" && (result === null || result === void 0 ? void 0 : result["Status"]) === api_1.GalaChainResponseType.Success) {
                await ctx.stub.flushWrites();
            }
            (_a = ctx === null || ctx === void 0 ? void 0 : ctx.logger) === null || _a === void 0 ? void 0 : _a.logTimeline("End Transaction", (_c = (_b = ctx.stub.getFunctionAndParameters()) === null || _b === void 0 ? void 0 : _b.fcn) !== null && _c !== void 0 ? _c : this.getName(), [{ chaincodeResult: result }]);
        });
    }
    async GetChaincodeVersion(ctx) {
        return api_1.GalaChainResponse.Success(this.version);
    }
    async GetContractAPI(ctx) {
        const methods = (0, GalaContractApi_1.getApiMethods)(this);
        const contractName = this.getName();
        return api_1.GalaChainResponse.Success({ contractName, methods, contractVersion: this.version });
    }
    GetObjectByKey(ctx, dto) {
        return api_1.GalaChainResponse.Wrap((0, utils_1.getPlainObjectByKey)(ctx, dto.objectId));
    }
    GetObjectHistory(ctx, dto) {
        return api_1.GalaChainResponse.Wrap((0, utils_1.getObjectHistory)(ctx, dto.objectId));
    }
}
exports.GalaContract = GalaContract;
tslib_1.__decorate([
    (0, GalaTransaction_1.GalaTransaction)({
        type: GalaTransaction_1.EVALUATE,
        out: "string"
    })
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    ,
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [types_1.GalaChainContext]),
    tslib_1.__metadata("design:returntype", Promise)
], GalaContract.prototype, "GetChaincodeVersion", null);
tslib_1.__decorate([
    (0, GalaTransaction_1.GalaTransaction)({
        type: GalaTransaction_1.EVALUATE,
        out: "object"
    })
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    ,
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [types_1.GalaChainContext]),
    tslib_1.__metadata("design:returntype", Promise)
], GalaContract.prototype, "GetContractAPI", null);
tslib_1.__decorate([
    (0, GalaTransaction_1.GalaTransaction)({
        type: GalaTransaction_1.EVALUATE,
        in: api_1.GetObjectDto,
        out: "object"
    }),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [types_1.GalaChainContext,
        api_1.GetObjectDto]),
    tslib_1.__metadata("design:returntype", Promise)
], GalaContract.prototype, "GetObjectByKey", null);
tslib_1.__decorate([
    (0, GalaTransaction_1.GalaTransaction)({
        type: GalaTransaction_1.EVALUATE,
        in: api_1.GetObjectHistoryDto,
        out: "object"
    }),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [types_1.GalaChainContext,
        api_1.GetObjectHistoryDto]),
    tslib_1.__metadata("design:returntype", Promise)
], GalaContract.prototype, "GetObjectHistory", null);
