// @gala-chain/chaincode/lib/src/mint/mintTokenWithAllowance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintTokenWithAllowance = void 0;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const api_1 = require("@gala-chain/api");
const bignumber_js_1 = require("bignumber.js");
const class_transformer_1 = require("class-transformer");
const allowances_1 = require("../allowances");
const mintToken_1 = require("./mintToken");
async function mintTokenWithAllowance(ctx, dto) {
    const { tokenClassKey, tokenInstance, quantity, owner } = dto;
    const response = await (0, allowances_1.grantAllowance)(ctx, {
        tokenInstance: (0, class_transformer_1.plainToInstance)(api_1.TokenInstanceQueryKey, { ...tokenClassKey, instance: tokenInstance }),
        allowanceType: api_1.AllowanceType.Mint,
        quantities: [{ user: ctx.callingUser, quantity }],
        uses: new bignumber_js_1.BigNumber("1"),
        expires: 0
    });
    const tokenInstanceArray = await (0, mintToken_1.mintToken)(ctx, {
        tokenClassKey: tokenClassKey,
        owner: owner,
        applicableAllowanceKey: (0, class_transformer_1.plainToInstance)(api_1.AllowanceKey, response[0]),
        quantity,
        authorizedOnBehalf: undefined
    });
    return tokenInstanceArray;
}
exports.mintTokenWithAllowance = mintTokenWithAllowance;
