// @gala-chain/chaincode/lib/src/token/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenAlreadyExistsError = exports.NftInstanceAllowanceMismatchError = exports.TokenClassNotFoundError = exports.NftDecimalError = exports.NotATokenAuthorityError = exports.InvalidDecimalError = exports.fetchTokenClassesWithPagination = exports.fetchTokenClasses = exports.fetchTokenClass = exports.updateTokenClass = exports.createTokenClass = exports.fetchTokenInstances = exports.fetchTokenInstance = void 0;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const TokenError_1 = require("./TokenError");
Object.defineProperty(exports, "InvalidDecimalError", { enumerable: true, get: function () { return TokenError_1.InvalidDecimalError; } });
Object.defineProperty(exports, "NftDecimalError", { enumerable: true, get: function () { return TokenError_1.NftDecimalError; } });
Object.defineProperty(exports, "NftInstanceAllowanceMismatchError", { enumerable: true, get: function () { return TokenError_1.NftInstanceAllowanceMismatchError; } });
Object.defineProperty(exports, "NotATokenAuthorityError", { enumerable: true, get: function () { return TokenError_1.NotATokenAuthorityError; } });
Object.defineProperty(exports, "TokenAlreadyExistsError", { enumerable: true, get: function () { return TokenError_1.TokenAlreadyExistsError; } });
Object.defineProperty(exports, "TokenClassNotFoundError", { enumerable: true, get: function () { return TokenError_1.TokenClassNotFoundError; } });
const createTokenClass_1 = require("./createTokenClass");
Object.defineProperty(exports, "createTokenClass", { enumerable: true, get: function () { return createTokenClass_1.createTokenClass; } });
const fetchTokenClasses_1 = require("./fetchTokenClasses");
Object.defineProperty(exports, "fetchTokenClass", { enumerable: true, get: function () { return fetchTokenClasses_1.fetchTokenClass; } });
Object.defineProperty(exports, "fetchTokenClasses", { enumerable: true, get: function () { return fetchTokenClasses_1.fetchTokenClasses; } });
Object.defineProperty(exports, "fetchTokenClassesWithPagination", { enumerable: true, get: function () { return fetchTokenClasses_1.fetchTokenClassesWithPagination; } });
const fetchTokenInstance_1 = require("./fetchTokenInstance");
Object.defineProperty(exports, "fetchTokenInstance", { enumerable: true, get: function () { return fetchTokenInstance_1.fetchTokenInstance; } });
Object.defineProperty(exports, "fetchTokenInstances", { enumerable: true, get: function () { return fetchTokenInstance_1.fetchTokenInstances; } });
const updateTokenClass_1 = require("./updateTokenClass");
Object.defineProperty(exports, "updateTokenClass", { enumerable: true, get: function () { return updateTokenClass_1.updateTokenClass; } });
